src/Hamiltonians/ho-cart-tools.jl

```diff
@@ -0,0 +1,203 @@
+        push!(df, (; block_id, block_E0, block_size = length(block_basis), addr, indices = t, t_basis))
+        !isnothing(save_to_file) && save_df(save_to_file, df)
+        if !isnothing(max_blocks) && block_id ≥ max_blocks
+            break
+        end
+    end
+    return df
+end
+
+# old version - issues with GC due to allocating many small vectors
+function get_all_blocks_comb(h; 
+        target_energy = nothing, 
+        max_energy = nothing, 
+        max_blocks = nothing, 
+        save_to_file = nothing,
+        kwargs...
+    )
+    add0 = starting_address(h)
+    N = num_particles(add0)
+    P = prod(h.S)
+
+    # initialise
+    df = DataFrame()
+    block_id = 0
+    known_basis = Set{typeof(add0)}()
+    tuples = with_replacement_combinations(1:P, N)
+    for t in tuples
+        # check target energy
+        block_E0 = noninteracting_energy(h, t)
+        if !isnothing(target_energy)
+            !isapprox(block_E0, target_energy; kwargs...) && continue
+        end
+        if !isnothing(max_energy)
+            block_E0 > max_energy && continue
+        end
+        # check if known
+        addr = BoseFS(P, t .=> ones(Int, N))
+        if addr in known_basis
+            continue
+        end
+
+        # new block found
+        block_id += 1
+        t_basis = @elapsed block_basis = build_basis(h, addr)
+        for b in block_basis
+            push!(known_basis, b)
+        end
+        push!(df, (; block_id, block_E0, block_size = length(block_basis), addr, indices = tuple(t...), t_basis))
+        !isnothing(save_to_file) && save_df(save_to_file, df)
+        if !isnothing(max_blocks) && block_id ≥ max_blocks
+            break
+        end
+    end
+    return df
+end
+
+"""
+    fock_to_cart(addr, S; zero_index = true)
+
+Convert a Fock state address `addr` to Cartesian harmonic oscillator basis 
+indices ``n_x,n_y,\\ldots``. These indices are bounded by `S` which is a 
+tuple of the maximum number of states in each dimension. By default the 
+groundstate in each dimension is indexed by `0`, but this can be changed 
+by setting `zero_index = false`.
+"""
+function fock_to_cart(
+    addr::SingleComponentFockAddress{N}, 
+    S::NTuple{D,Int}; 
+    zero_index = true
+    ) where {N,D}
+    prod(S) == num_modes(addr) || throw(ArgumentError("Specified cartesian states are incompatible with address"))
+    states = CartesianIndices(S)
+
+    cart = vcat(map(
+        p -> [Tuple(states[p.mode]) .- Int(zero_index) for _ in 1:p.occnum], 
+        OccupiedModeMap(addr))...)
+
+    return SVector{N,NTuple{D,Int}}(cart)
+end
+
+"""
+    occupied_modes_list(addr)
+
+Output a `Tuple` of all occupied modes in `addr`. Multiply-occupied modes 
+are listed as duplicates so the length of the output is always the number 
+of particles. This is a compact form of [`onr`](@ref) when number of particles
+is less than the umber of modes.
+"""
+@inline function occupied_modes_list(a::SingleComponentFockAddress{N}) where {N}
```

Comment:
I can't find this function used anywhere, but it is exported. Is it used as part of the workflow? It could be rewritten to have no allocations, but if it's not meant to be used in a tight loop, I'd say it's fine.

Reply:
I removed it since it is not really needed